src/3rd_party/topk.cuh

```diff
@@ -0,0 +1,302 @@
+
+  for(int ite = 0; ite < k; ite++) {
+    partial.init(descendingOrder);
+    const IndexType threadStart = tid + block_lane * BLOCK_SIZE_;
+
+    // This is needed to ensure the indices for the threads in each valid block starts in a valid range for that block.
+    if(threadStart < vocab_size) partial.p = threadStart;
+    #pragma unroll
+    for(IndexType elem_id = threadStart; elem_id < vocab_size; elem_id += BLOCK_SIZE_ * BLOCKS_PER_BEAM_) {
+      IndexType index = elem_id + tmp_log_buf_index;
+      descendingOrder? partial.insertKeepMax(log_probs[index], index) : partial.insertKeepMin(log_probs[index], index);
+    }
+
+    TopK<IndexType, T> total = BlockReduce(temp_storage).Reduce(partial, descendingOrder? reduce_topk_max<IndexType, T>: reduce_topk_min<IndexType, T>);
+
+    if (tid == 0) {
+      const int index = tmp_topk_buf_index + ite;
+      topk_tmp_id_buf[index] = getRowOffsets? total.p - tmp_log_buf_index : total.p;
+      topk_tmp_val_buf[index] = total.u;
+      // If we found a max, blank out the value in the log prob array before starting the next iteration.
+      // Otherwise, we don't need to issue a write since all prob values must have been T::min()
+      if(total.u != minimal) log_probs[total.p] = minimal;
+    }
+    __syncthreads();
+  }
+
+  // Update prob array with original values.
+  for(int beam = tid; beam < k; beam += BLOCK_SIZE_) {
+    const IndexType index = tmp_topk_buf_index + beam;
+    T val = topk_tmp_val_buf[index];
+    // We only want to replace the value in the log prob array if a value was blanked out (we found a max).
+    // When a max isn't found, topk_tmp_val_buf[index] will be T::min()
+    if(val != minimal) {
+      IndexType k_idx = getRowOffsets? topk_tmp_id_buf[index] + tmp_log_buf_index : topk_tmp_id_buf[index];
+      log_probs[k_idx] = (T)topk_tmp_val_buf[index];
+    } 
+  }
+}
+
+template<typename IndexType, typename T, int BLOCK_SIZE_, int BLOCKS_PER_BEAM_>
+__global__ void topk_stage_2(const IndexType* __restrict topk_tmp_id_buf,
+                             T* topk_tmp_val_buf,
+                             TopK<IndexType, T>* top,
+                             IndexType* outIndices,
+                             T* outVals,
+                             const int beams_per_batch,
+                             const int k,
+                             bool descendingOrder) {
+
+  const int size = beams_per_batch * k * BLOCKS_PER_BEAM_; 
+  const int tid = threadIdx.x;
+  const int batch_id = blockIdx.x;
+  const T minimal = descendingOrder? -FpInfinity<T>::infinity() : FpInfinity<T>::infinity();;
+
+  typedef cub::BlockReduce<TopK<IndexType, T>, BLOCK_SIZE_> BlockReduce;
+  __shared__ typename BlockReduce::TempStorage temp_storage;
+  extern __shared__ char array[];
+  T *s_val = topk_tmp_val_buf + batch_id * size;
+  TopK<IndexType, T> *topks = (TopK<IndexType, T>*)(array);
+  
+  TopK<IndexType, T> partial;
+
+  for(int ite = 0; ite < k; ite++) {
```

Comment:
Would you mind adding a headline comment to each of the sections in this function to summarize in a single line what they achieve? It is faster to read a single line than having to parse the code and understand the logic. Thanks!